src/encoder.rs

```diff
@@ -668,27 +668,67 @@ impl<W: Write> Writer<W> {
 
         let prev = vec![0; in_len];
         let mut prev = prev.as_slice();
-        let mut current = vec![0; in_len];
 
-        let mut zlib = ZlibEncoder::new(Vec::new(), self.info.compression.to_options());
         let bpp = self.info.bpp_in_prediction();
         let filter_method = self.options.filter;
         let adaptive_method = self.options.adaptive_filter;
 
-        for line in data.chunks(in_len) {
-            let filter_type = filter(
-                filter_method,
-                adaptive_method,
-                bpp,
-                prev,
-                &line,
-                &mut current,
-            );
-            zlib.write_all(&[filter_type as u8])?;
-            zlib.write_all(&current)?;
-            prev = line;
-        }
-        let zlib_encoded = zlib.finish()?;
+        let zlib_encoded = match self.info.compression {
+            Compression::Fast => {
+                let mut compressor = fdeflate::Compressor::new(std::io::Cursor::new(Vec::new()))?;
+
+                let mut current = vec![0; in_len + 1];
+                for line in data.chunks(in_len) {
+                    let filter_type = filter(
+                        filter_method,
+                        adaptive_method,
+                        bpp,
+                        prev,
+                        line,
+                        &mut current[1..],
+                    );
+
+                    current[0] = filter_type as u8;
+                    compressor.write_data(&current)?;
+                    prev = line;
+                }
+
+                let compressed = compressor.finish()?.into_inner();
+                if compressed.len()
+                    > fdeflate::StoredOnlyCompressor::<()>::compressed_size((in_len + 1) * height)
+                {
+                    let mut compressor =
+                        fdeflate::StoredOnlyCompressor::new(std::io::Cursor::new(Vec::new()))?;
+                    for line in data.chunks(in_len) {
+                        compressor.write_data(&[0])?;
```

Comment:
This hard-codes writing unfiltered image data even if the caller requested a filter. Given that the data is being written uncompressed, doing filtering again would only add performance cost (both for encoding _and_ subsequent decoding), but seems worth mentioning

Reply:
This sounds like it would make a good code comment - could you add it?